### ISSUE TITLE
Support HTTPS

### ISSUE THREAD

**Issue by juntao** [WasmEdge/WasmEdge#1530]
## Motivation

The WasmEdge WASI socket API supports HTTP networking in Wasm apps. However, it currently lacks HTTPS and TLS support.

While work is being done to support TLS through wasi-crypto, we cannot wait for it. Here, we propose to support a limited set of HTTPS operations directly via a WasmEdge plugin. 

## Details

* Create a WasmEdge plugin that supports HTTPS GET and POST requests via C++ based host functions.
* Expose those host functions as a  Rust API for developers. We should just add to the [wasmedge_http_req](https://github.com/second-state/wasmedge_http_req) crate. It is okay for the requests to block.

**Comment by 2019zhou** [comment]
Plz assign this to me

**Comment by 2019zhou** [comment]
## Progress

- [x]  set up the work environment,
- [x]  Understand the Wasmedge Plugin API
- [x]  Understand the existing implementation of HTTP API and figure out the difference between implementing HTTP API and HTTPS API
- [x]  Read through the wasmedge_http_req
- [ ]  Implement the ssl supported C++ host function https://github.com/WasmEdge/WasmEdge/
- [ ]  Add supported C++ host function to the https://github.com/second-state/wasmedge_http_req


    
### Understand the Wasmedge Plugin API

Wasmedge plugin API enables contributing new functions to the Wasmedge Runtime. [[Host functions](https://github.com/WasmEdge/WasmEdge/issues/%5B%3Chttps://wasmedge.org/book/en/extend/plugin/hostfunction.html%3E%5D(%3Chttps://wasmedge.org/book/en/extend/plugin/hostfunction.html%3E))]([https://wasmedge.org/book/en/extend/plugin/hostfunction.html](https://wasmedge.org/book/en/extend/plugin/hostfunction.html)) is a recommended way of implementing plugin API. As a matter of fact, wasmedge has several plugin API implemented in a host function. 

e.g. 

support for TensorFlow

[https://github.com/second-state/WasmEdge-tensorflow](https://github.com/second-state/WasmEdge-tensorflow)

support for key-value storage

[https://github.com/second-state/WasmEdge-storage](https://github.com/second-state/WasmEdge-storage)

support for image

[https://github.com/second-state/WasmEdge-image](https://github.com/second-state/WasmEdge-image)

With the goal of creating a new plugin API in a host function way, we first need to create the host function definitions. Then we finish the host function implementations. Finally, Register the implementation in CMakeLists and the example module in the CLI tool. 

The following link gives a concrete example of doing the steps to create the host function.

[https://github.com/WasmEdge/WasmEdge/compare/master...challenge/proposal](https://github.com/WasmEdge/WasmEdge/compare/master...challenge/proposal)

that is, the host modules need to be defined and implemented and the host module needs to be registered to WasmEdge.  

### Understand the existing implementation HTTP API and figure out the difference between implementing HTTP API and HTTPS API**

- existing HTTP API

[https://github.com/second-state/wasmedge_http_req](https://github.com/second-state/wasmedge_http_req)

wasmedge_http_req

learn the API for HTTP GET and POST

It can be noticed that the API is called wasmedge_http_req::requests::get, wasmedge_http_req::requests::post

```python
use wasmedge_http_req::request;

fn main() {
    let mut writer = Vec::new(); //container for body of a response
    let res = request::get("http://127.0.0.1/", &mut writer).unwrap();

    println!("Status: {} {}", res.status_code(), res.reason());
}

fn main() {
    let mut writer = Vec::new(); //container for body of a response
    const BODY: &[u8; 27] = b"field1=value1&field2=value2";
    let res = request::post("https://httpbin.org/post", BODY, &mut writer).unwrap();

    println!("Status: {} {}", res.status_code(), res.reason());
    println!("Headers {}", res.headers());
    //println!("{}", String::from_utf8_lossy(&writer));
}
```

Note that the original repositoryhttps://github.com/jayjamesjay/http_req has realized the built-in HTTPS support. According to 
[https://github.com/WasmEdge/WasmEdge/issues/1430](https://github.com/WasmEdge/WasmEdge/issues/1430)
 Rustls is based on the Ring library, which cannot be compiled into WebAssembly. Therefore, we are unable to directly use the Rusttls. 

The requirement is that the API for calling HTTPS and HTTP are the same. Thus, it is preferable to directly implement a C++ module using OpenSSL to replace the rusttls.

**Comment by q82419** [comment]
For the plugins, you can refer to:

1. The [plugin example](https://github.com/WasmEdge/WasmEdge/tree/master/examples/plugin/get-string) . 
2. The [wasmedge_process plugin](https://github.com/WasmEdge/WasmEdge/tree/master/plugins/wasmedge_process) and [its unit tests](https://github.com/WasmEdge/WasmEdge/tree/master/test/plugins/wasmedge_process) .

Thanks.

**Comment by stuburger** [comment]
Hi, does this mean that we can start calling https enabled endpoints from [wasmedge-quickjs](https://github.com/second-state/wasmedge-quickjs) programs? I found this after I saw that my requests were falling back to http and failing. This looks like an important and exciting piece of work.

**Comment by 2019zhou** [comment]
It implements the host functions for [https request for Rust programs]( https://github.com/second-state/wasmedge_http_req) .
And I think the host function for https request might also be used for quickjs programs. But I have not learned about the wasmedge-quickjs program yet. 
@q82419

**Comment by juntao** [comment]
Yes, once this is done, we will be able to make HTTPS requests from QuickJS. The big limitation for this approach is that it only supports blocking requests.

To truly support HTTPS async networking in QuickJS, we will need to wait for this issue:

https://github.com/WasmEdge/WasmEdge/issues/1549

> Hi, does this mean that we can start calling https enabled endpoints from [wasmedge-quickjs](https://github.com/second-state/wasmedge-quickjs) programs? I found this after I saw that my requests were falling back to http and failing. This looks like an important and exciting piece of work.

**Comment by 2019zhou** [comment]
### problems
1. wasmedge host function cannot pass the string. If Wasmedge host function can pass the string, it can use the exact same API and by adding the few lines of Rust code, that is res.status_code(), res.headers(), res.reason() is not available().
```
let mut resvec = Vec::new();
let  res = Response::try_from(RESPONSE, &mut resvec).unwrap();
return Ok(res);
```
Now, it can only print all the feedback by calling post/get/head.

2. Is there any need to add more functions other than the [send_data function](https://github.com/2019zhou/wasmedge_http_req/blob/zhou/httpsreq/src/httpsreq.rs#L7)
Because WASM built-in types (aka. uint32_t, uint64_t, float, and double) are supported in host functions, it is possible to add a get_status function.

**Comment by hydai** [comment]
For question 1, you can pre-allocate the space for a string in wasm and then pass its pointer and length to the host function, and then load/store the string into the space. You can check how we handle the `getaddrinfo` here: https://github.com/WasmEdge/WasmEdge/blob/master/lib/host/wasi/inode-linux.cpp#L793

**Comment by 2019zhou** [comment]
Sorry for the unclear explanations. 
The string can be passed into the Httpsreq https://github.com/2019zhou/WasmEdge/blob/zhou/httpsreq/plugins/httpsreq/httpsreqfunc.cpp#L21
but can not be returned or passed out the Host function.
In order to achieve the goal of using the exact same API as HTTP request, the string needs to be returned.

**Comment by 2019zhou** [comment]
It looks like the string can be passed out of the host function https://github.com/WasmEdge/WasmEdge/blob/master/plugins/wasmedge_process/processfunc.cpp#L324-L340.
I am learning and implementing this part.

**Comment by hydai** [comment]
Hi @juntao 
Since we already have rustls support, is this issue still needed?

**Comment by hydai** [comment]
The HTTPS feature should be supported by rusttls (compiled into WASM). Closing this issue; please feel free to reopen it if there is anything still needing implementation.
